gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagementStateStore.java

```diff
@@ -0,0 +1,167 @@
+   * @param dagId DagId of the dag
+   */
+  boolean containsDag(DagManager.DagId dagId) throws IOException;
+
+  /**
+   * Returns a dag if present, null otherwise.
+   * @param dagId DagId of the dag
+   */
+  Dag<JobExecutionPlan> getDag(DagManager.DagId dagId) throws IOException;
+
+  /**
+   * Delete the {@link Dag} from the backing store, typically called upon completion of execution.
+   * @param dag The dag completed/cancelled execution on {@link org.apache.gobblin.runtime.api.SpecExecutor}.
+   */
+  default void deleteDag(Dag<JobExecutionPlan> dag) throws IOException {
+    deleteDag(DagManagerUtils.generateDagId(dag));
+  }
+
+  /**
+   * Delete the {@link Dag} from the backing store, typically upon completion of execution.
+   * @param dagId The ID of the dag to clean up.
+   */
+  void deleteDag(DagManager.DagId dagId) throws IOException;
+
+  /**
+   * This marks the dag as a failed one.
+   * Failed dags are queried using {@link DagManagementStateStore#getFailedDagIds()} later to be retried.
+   * @param dag failing dag
+   * @throws IOException
+   */
+  void markDagFailed(Dag<JobExecutionPlan> dag) throws IOException;
+
+  /**
+   * Return a list of all failed dags' IDs contained in the dag state store.
+   */
+  Set<String> getFailedDagIds() throws IOException;
+
+  /**
+   * Returns the failed dag.
+   * If the dag is not found or is not marked as failed through {@link DagManagementStateStore#markDagFailed(Dag)}, it returns null.
+   * @param dagId dag id of the failed dag
+   */
+  Dag<JobExecutionPlan> getFailedDag(DagManager.DagId dagId) throws IOException;
+
+  /**
+   * Deletes the failed dag. No-op if dag is not found or is not marked as failed.
+   * @param dag
+   * @throws IOException
+   */
+  default void deleteFailedDag(Dag<JobExecutionPlan> dag) throws IOException {
+    deleteFailedDag(DagManagerUtils.generateDagId(dag));
+  }
+
+  void deleteFailedDag(DagManager.DagId dagId) throws IOException;
+
+  /**
+   * Adds state of a {@link org.apache.gobblin.service.modules.flowgraph.Dag.DagNode} to the store.
+   * Note that a DagNode is a part of a Dag and may already be present in the store thorugh
+   * {@link DagManagementStateStore#checkpointDag}. This call is just an additional identifier which may be used
+   * for DagNode level operations. In the future, it may be merged with checkpointDag.
+   * @param dagNode dag node to be added
+   * @param dagId dag id of the dag this dag node belongs to
+   * @throws IOException
+   */
+  void addDagNodeState(Dag.DagNode<JobExecutionPlan> dagNode, DagManager.DagId dagId)
+      throws IOException;
+
+  /**
+   * Returns the requested {@link  org.apache.gobblin.service.modules.flowgraph.Dag.DagNode}
+   * @param dagNodeId of the dag ndoe
+   */
+  Dag.DagNode<JobExecutionPlan> getDagNode(DagNodeId dagNodeId);
+
+  /**
+   * Returns a list of {@link org.apache.gobblin.service.modules.flowgraph.Dag.DagNode} for a {@link Dag}
```

Comment:
not documented here, but the impl seems to return an empty list when not found.  I don't believe a real Dag could ever have no DagNodes, so, yes, that could potentially be a sentinel value for "not found"... but overall, we probably want to more explicitly communicate an unknown `DagId` to handle that specifically

Reply:
updating javadoc